File: Q2683_「ロング・ロング・ストリング」問題/answer.ml
(* Try CodeIQ                                 *)
(* author: Leonardone @ NEETSDKASU            *)
(* ****************************************** *)

let f n = 1.0 +. (floor (n *. (log10 n))) in

let g m =
	let rec h b t =
		if b > t then -1.0
		else
			let x = floor ((b +. t) /. 2.0) in
			let y = f x in
			if y < m then h (x +. 1.0) t
			else if y > m then h b (x -. 1.0)
			else x
	in h 2.0 (10.0 ** 10.0) in
				
let m = read_float () in
let ans = Int64.to_string (Int64.of_float (g m)) in

print_newline (print_string ans)

;;

File: Q2683_「ロング・ロング・ストリング」問題/answer.nim
# Try CodeIQ
# author: Leonardone @ NEETSDKASU

import parseutils
import math

proc flr(x: float): float = return x.int64.float

proc f(n: float): float = return 1.0 + flr(n * log10(n))

proc h(b, t, m: float): float =
    if b > t:
        return -1.0
    else:
        var x = flr((b + t) / 2.0)
        var y = f(x)
        if y < m:
            return h(x + 1.0, t, m)
        elif y > m:
            return h(b, x - 1.0, m)
        else:
            return x

proc g(m: float): float = return h(2.0, pow(10.0, 10.0), m)

var m = 0.0
var r = parseFloat(readLine(stdin), m)
var ans = g(m).int64
echo(ans)
